### PR TITLE
Add console.reactStack() call to expose warnings to react-error-overlay

### PR DESCRIPTION
**Pull request by sw-yx**
as discussed with Dan - context from him below

> I recently made it so that warning() calls automatically attach the stack. It’s available right in that module https://github.com/facebook/react/blob/master/packages/shared/warning.js

> Now that we always have stack info, we can expose it to other tools consistently too

> In fact Create React App’s error overlay already intercepts console.error calls if a special method called console.reactStack was called before

> React 15 made console.reactStack calls only for one warning, to test the idea. We never implemented it for other warnings and then didn’t implement it at all in 16

> Suggestion: take a look at the API that React 15 was calling (search the UMD bundle for console.reactStack calls). Try doing the same in 16, except now in the centralised place (warning module). Use Create React App to check if it worked. When it works it should start showing the error overlay for warnings too, and include component stack as actual frames in the overlay

---

# CRA test results (these warnings are what we want):

example of using the new exposed `ReactDebugCurrentFrame.currentFiber`:

image

example of using adding the `currentlyValidatingElement` on top of the frames from `ReactDebugCurrentFrame.getStackFrames` (otherwise this would show at `<Test />`):

image

In future we may modify react-error-overlay to better show warnings (prior issue: https://github.com/facebook/create-react-app/issues/1888) but this lays the groundwork to do that

### CODE REVIEW

**Comment by react-linter** [comment]
<details>
  <summary>Details of bundled changes.</summary>

  <p>Comparing: 2c560cb9952c05fb0a45aed87723c9a8bf6ff8d4...e072e0ef5b17ad8fc295b6f1794db84656245777</p>


  
## react
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react.development.js | +2.9% | +2.3% | 57.39 KB | 59.08 KB | 16.03 KB | 16.4 KB | UMD_DEV
react.development.js | +3.3% | +2.7% | 51.58 KB | 53.28 KB | 14.21 KB | 14.58 KB | NODE_DEV
React-dev.js | +2.2% | +1.6% | 49.87 KB | 50.95 KB | 13.78 KB | 14 KB | FB_WWW_DEV

## react-dom
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-dom.development.js | +0.1% | +0.1% | 639.44 KB | 640.1 KB | 149.88 KB | 150.07 KB | UMD_DEV
react-dom.development.js | +0.1% | +0.1% | 635.57 KB | 636.24 KB | 148.7 KB | 148.9 KB | NODE_DEV
react-dom-server.browser.development.js | +0.6% | +0.6% | 102.12 KB | 102.73 KB | 27.21 KB | 27.38 KB | UMD_DEV
react-dom-server.browser.development.js | +0.6% | +0.6% | 98.25 KB | 98.86 KB | 26.28 KB | 26.44 KB | NODE_DEV
react-dom-server.node.development.js | +0.6% | +0.6% | 100.17 KB | 100.79 KB | 26.81 KB | 26.97 KB | NODE_DEV
ReactDOM-dev.js | 0.0% | 0.0% | 642.5 KB | 642.55 KB | 147.18 KB | 147.2 KB | FB_WWW_DEV

## react-art
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-art.development.js | +0.2% | +0.2% | 425.93 KB | 426.59 KB | 96.09 KB | 96.32 KB | UMD_DEV
react-art.development.js | +0.2% | +0.3% | 358.49 KB | 359.15 KB | 79.02 KB | 79.25 KB | NODE_DEV
ReactART-dev.js | 0.0% | 0.0% | 348.4 KB | 348.45 KB | 73.95 KB | 73.95 KB | FB_WWW_DEV

## react-test-renderer
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-test-renderer.development.js | +0.2% | +0.3% | 362.86 KB | 363.53 KB | 79.55 KB | 79.78 KB | UMD_DEV
react-test-renderer.development.js | +0.2% | +0.3% | 358.99 KB | 359.66 KB | 78.58 KB | 78.81 KB | NODE_DEV
ReactTestRenderer-dev.js | 0.0% | 0.0% | 363.87 KB | 363.92 KB | 77.59 KB | 77.6 KB | FB_WWW_DEV

## react-reconciler
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-reconciler.development.js | +0.2% | +0.3% | 346.88 KB | 347.54 KB | 74.92 KB | 75.15 KB | NODE_DEV
react-reconciler-persistent.development.js | +0.2% | +0.3% | 345.5 KB | 346.16 KB | 74.37 KB | 74.59 KB | NODE_DEV

## react-native-renderer
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
ReactNativeRenderer-dev.js | +0.1% | +0.2% | 480.32 KB | 481.01 KB | 106.03 KB | 106.27 KB | RN_FB_DEV
ReactNativeRenderer-dev.js | +0.1% | +0.2% | 480.05 KB | 480.74 KB | 105.97 KB | 106.21 KB | RN_OSS_DEV
ReactFabric-dev.js | +0.1% | +0.2% | 470.53 KB | 471.22 KB | 103.59 KB | 103.83 KB | RN_FB_DEV
ReactFabric-dev.js | +0.1% | +0.2% | 470.57 KB | 471.26 KB | 103.6 KB | 103.84 KB | RN_OSS_DEV

  </details>
  
<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

File: packages/react-reconciler/src/ReactCurrentFiber.js [hunk 1]
```diff
@@ -91,6 +91,7 @@ export function resetCurrentFiber() {
 export function setCurrentFiber(fiber: Fiber) {
   if (__DEV__) {
     ReactDebugCurrentFrame.getCurrentStack = getCurrentFiberStackInDev;
+    ReactDebugCurrentFrame.currentFiber = fiber;
```
Comment (gaearon) [inline packages/react-reconciler/src/ReactCurrentFiber.js]: We don't want to expose the fiber itself. This module is meant to work with different implementations too. For example server renderer sets its own `ReactDebugCurrentFrame.getCurrentStack` implementation that doesn't use Fibers internally.

I was thinking something more like 

```js
ReactDebugCurrentFrame.getCurrentStackFrames = ...
```

which returns the object. Set it together with `ReactDebugCurrentFrame.getCurrentStack`. Later we'll remove `ReactDebugCurrentFrame.getCurrentStack`.

**Comment by sw-yx** [comment]
ok i think i got it

> The "react" module shouldn't know about fibers.

btw i wondered about this too but then i found https://github.com/facebook/react/blob/3c977dea6b96f6a9bb39f09886848da870748441/packages/react/src/ReactCurrentOwner.js#L10 which makes the same reference I do. just calling it out in case its an issue, i'll move my implementation inside the renderer

File: packages/react-reconciler/src/ReactCurrentFiber.js (hunk 1 above)
Comment (sw-yx) [reply packages/react-reconciler/src/ReactCurrentFiber.js]: ok. i guess i shoudl have asked about this haha, i was also unsure about it. i think i understand.

**Comment by sw-yx** [comment]
DX question - i dont know where else to ask - is there a "build watch" mode or a "partial build" mode? right now the slowest part of this process is building the bundles from scratch every time (for testing in CRA) and it builds everything, right down to React Fabric which seems excessive. i'd be fine paying this cost once but it seems there is no alternative of incremental or partial builds?

**Comment by sw-yx** [comment]
I'm a bit stuck with the testing strategy:

1. not sure where to put it, it doesnt seem to fit with any of the existing test files. i made a separate test file for now under `ReactStackFrameWarnings-test.internal.js`
2. **For the `react-reconciler` test**, [I am basically mocking `console.reactStack` and expecting it to be called.](https://github.com/sw-yx/react/blob/addReactStackCall/packages/react-reconciler/src/__tests__/ReactStackFrameWarnings-test.internal.js) However it is not being called and I can't figure out why. **Do I need to be using a different renderer than `ReactNoop`?** I barely see any other tests that use, for example, `react-reconciler`.
3. **For the `ReactPartialRenderer` test**, I have yet to get to this. hoping that solving 1 and 2 will give me some insight into solving this one.

Not sure if this is supposed to be super obvious and I'm missing something..

File: packages/react-dom/src/server/ReactPartialRenderer.js [hunk 2]
```diff
@@ -152,6 +164,35 @@ if (__DEV__) {
     }
     return stack;
   };
+
+  getCurrentServerStackFramesImpl = function(): ReactFrame[] {
```
Comment (gaearon) [inline packages/react-dom/src/server/ReactPartialRenderer.js]: It seems like once you have this, you should be able to rewrite `getCurrentServerStackImpl` on top of it and remove some of the duplication between them.

File: packages/react-reconciler/src/ReactCurrentFiber.js [hunk 3]
```diff
@@ -80,9 +104,22 @@ export function getCurrentFiberStackInDev(): string {
   return '';
```
Comment (gaearon) [inline packages/react-reconciler/src/ReactCurrentFiber.js]: You can rewrite this on top of `getCurrentFiberStackFramesInDev` now

**Comment by gaearon** [comment]
>However it is not being called and I can't figure out why.

What is your methodology for figuring out why it didn't get called? What did you try?

File: packages/react-dom/src/server/ReactPartialRenderer.js (hunk 2 above)
Comment (sw-yx) [reply packages/react-dom/src/server/ReactPartialRenderer.js]: done

File: packages/react-reconciler/src/ReactCurrentFiber.js (hunk 3 above)
Comment (sw-yx) [reply packages/react-reconciler/src/ReactCurrentFiber.js]: after studying this a bit, I think this suggestion would cause additional duplication because i would need to pull in the work done by `getStackByFiberInDevAndProd` and there are other modules that depend on `getStackByFiberInDevAndProd`. So instead of doing a rewrite of `getCurrentFiberStackInDev` I will do a rewrite of `getStackByFiberInDevAndProd` and just get `getCurrentFiberStackInDev` to rely on it.

**Comment by gaearon** [comment]
>in the test that I wrote, i mocked console.reactStack and then tried to render a component that should have raised a warning:

Why do you expect it to be called when no warning has been raised? My understanding is that we're only calling it when there is a warning. You can try using a key warning, for example.

**Comment by sw-yx** [comment]
i think you might have misread me there. i -did- try something that should have raised a warning:

```js
function Foo() {
  return <div class="bar" />; // should raise className warning
}
```

i just took another run at this and i think my problem is still not having the correct testing strategy figured out. there are any number of ways to do this so maybe to help narrow down the potential solution space here are some binary 
questions that should be easy to answer:

1. **Location**: should i locate the test (for this reactStack call) in `packages/react-dom` or `packages/react-reconciler`? or does it not matter?
2. **Test package**: should i use `ReactTestUtils = require('react-dom/test-utils');` or `ReactNoop = require('react-noop-renderer');`?
3. **Spy/Mock strategy**: should I use `spyOnDev(console, 'reactStack');` or maybe a more traditional jest mock like `const mockStack = jest.fn(); console.reactStack = mockStack;`?
4. **What to test for**: I actually think I know this, but I might as well be doubly sure - my plan is test that 1) the `console.reactStack` call happened and 2)  `console.reactStack` is passed an array of stack frame objects. That's all I can think to test for!

there are a wide, wide variety of tests in the React codebase and not all of them seem to pass so I am quite uncertain what to try. Especially the ones that assume `document` exists (so I assume that's `jsdom` in the background) look promising but I'm not even sure if that is the right way because it doesn't work for me (I think I may need to dig deeper into how jsdom is set up in the jest config).

Sorry this is a bit wordy, I figured I would try my best to communicate how I am blocked. @nhunzaker if you have a half hour or so I could definitely pair with you on this one too.

**Comment by gaearon** [comment]
>Location: should i locate the test (for this reactStack call) in packages/react-dom or packages/react-reconciler? or does it not matter?

It doesn’t matter.

>Test package: should i use ReactTestUtils = require('react-dom/test-utils'); or ReactNoop = require('react-noop-renderer');

Noop renderer doesn’t have any DOM-specific warnings. So testing for `className` warning wouldn’t give you anything.

I’d just use plain ReactDOM API. You can use TestUtils too but they’re just a confusing wrapper over ReactDOM. 

>Spy/Mock strategy: should I use spyOnDev(console, 'reactStack'); or maybe a more traditional jest mock like const mockStack = jest.fn(); console.reactStack = mockStack;?

I would use a “traditional” mock.

>What to test for: I actually think I know this, but I might as well be doubly sure - my plan is test that 1) the console.reactStack call happened and 2) console.reactStack is passed an array of stack frame objects. That's all I can think to test for!

That sounds good. 

>there are a wide, wide variety of tests in the React codebase and not all of them seem to pass 

That seems like the root of the problem? I don’t know what you mean because all tests *should* pass. They do on CI. So if they don’t work for you we’ll need more details about that. What doesn’t pass? What are you seeing?
